Add io.github.rafostar.Clapper.Enhancers

Rafostar opened the pull request:
<!-- ⚠️⚠️ Submission pull request MUST be made against the `new-pr` **base branch** ⚠️⚠️  -->

<!-- 💡 Go to the preview tab to click the links below 💡 -->

### Please confirm your submission meets all the criteria

<!-- 💡 Please replace each `[ ]` with `[X]` when the step is complete 💡 -->

- [x] Please describe the application briefly.

This is an **extension** submission for `com.github.rafostar.Clapper` (I am using new `io.github` ID for this  as required).

In latest stable release, [Clapper](https://flathub.org/apps/com.github.rafostar.Clapper) has an ability to load custom `libpeas` plugins for which term "enhancers" is used. This extension would contain the set of default ones - made by the author of the Clapper app (me :smile:). Currently they are built as part of Clapper manifest, but I would like to move them into separate extension with a mount endpoint, for reasons:

* Besides Clapper itself, there are already 3 other apps in Flathub that want to use this (tubefeeder, tuba, newsflash)
* Allow installing them (plugins) as an option

- [x] The domain used for the application ID is [controlled by the application developer(s)][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read and followed all the [Submission requirements][reqs] and the [Submission guide][reqs2].
- [x] I have [built][build] and tested the submission locally.

I have built and tested this extension in both Clapper app (which is pending corresponding changes in this [PR](https://github.com/flathub/com.github.rafostar.Clapper/pull/35)) and also tested as an extension to another app by using modified manifest for testing purposes.

- [x] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. [Link](https://github.com/rafostar/clapper-enhancers)

This PR also includes `README.md` file with description how to implement this extension for another application.

<!-- ⚠️⚠️ DO NOT modify anything below this line ⚠️⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[reqs2]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented:
why do this need to be an extension?

Rafostar commented:
> why do this need to be an extension?

Mainly for the reasons I have written in this PR description. At least 3 other apps (from different developers) want to use this, but aren't too happy with a chore of bundling (and maintaining) dependency of a plugin like `yt-dlp` which needs constant updates in sync with plugin code itself. They are also concerned with their app size increment, so we figured to bundle this as an extension that apps can share and maintain updates in a single repo to make life easier.

Plugins built here are dynamically loaded by apps if/when needed, so I hope this fits what an extension is supposed to do.

hfiguiere commented:
The problem with N:1 plugins is that they will usually come with a lot of FRAGILE dependencies,. yt-dlp for example seems to need python. The python version is tied to the runtime for example.
And you are opening the door to conflicts with python modules from various places.

This is open another boatload of problems for very little benefits.

Rafostar commented:
> And you are opening the door to conflicts with python modules from various places.
> The python version is tied to the runtime for example.
> This is open another boatload of problems for very little benefits.

OK. Thanks for confirming.

Maintaining different branches like `branch/24.08` would probably partially work around this problem, but doing so would likely put more effort into maintenance (especially with backporting) than benefit. Even more problems if some of the dependencies change over time or are added. Even with branches, there would be update restrictions between plugin functionality (its available public methods) and the app that uses it, so it would be another problem in the boilerplate.

So my understanding is that you recommend simply packaging additional modules/deps in their manifests instead of doing shared extension to avoid collisions. And to ensure they use latest working versions, just use `x-data-checker` in them, yes?

bbhtt commented on io.github.rafostar.Clapper.Enhancers.Default.json:
```diff
@@ -0,0 +1,68 @@
+    "sdk": "org.freedesktop.Sdk//24.08",
+    "build-extension": true,
+    "separate-locales": false,
+    "appstream-compose": true,
```
```suggestion
```

this is default

Rafostar replied:
Done

bbhtt commented on io.github.rafostar.Clapper.Enhancers.Default.json:
```diff
@@ -0,0 +1,68 @@
+                }
+            ]
+        },
+        {
+            "name": "metainfo",
+            "buildsystem": "simple",
+            "sources": [
+                {
+                    "type": "file",
+                    "path": "io.github.rafostar.Clapper.Enhancers.Default.metainfo.xml"
+                }
+            ],
+            "build-commands": [
+                "install -Dm644 -t ${FLATPAK_DEST}/share/metainfo ${FLATPAK_ID}.metainfo.xml"
+            ]
+        }
```
merge the sources with the above and add build-commands as `post-install`.

Rafostar replied:
I did that, but I also had to add `../` to file path. It seems `meson` build system runs inside build dir and not in project root.

bbhtt commented on io.github.rafostar.Clapper.Enhancers.Default.metainfo.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="addon">
+  <id>io.github.rafostar.Clapper.Enhancers.Default</id>
+  <extends>io.github.rafostar.Clapper.Enhancers</extends>
```
It extends the app, extending an extension point makes no sense. `com.github.rafostar.Clapper`

Rafostar replied:
OK, renamed `io.github.rafostar.Clapper.Enhancers.Default` -> `io.github.rafostar.Clapper.Enhancers` and made it simply extend Clapper app.

bbhtt commented:
I don't see the extension point for this in Clapper.

bbhtt commented:
> So my understanding is that you recommend simply packaging additional modules/deps in other apps manifests instead of doing shared extension to avoid collisions. And to ensure they use latest working versions, just use x-data-checker in them, yes?

Yea this seems better than another extension. Unless the module is very large in size or complicated to maintain with the app or legal issues like patented codecs, creating an extension is not worth it.

Rafostar commented:
> I don't see the extension point for this in Clapper.

I did a PR for that here: https://github.com/flathub/com.github.rafostar.Clapper/pull/35

Just waiting to see whether this will be green-lit before merging (also, possibly adjusting it with the changes that might be requested here).

Rafostar commented:
As requested in review, I dropped `.Default` suffix and made this simply extend app. Renamed all files, updated readme and PR name.

> Unless the module is very large in size

Well, depends on what the definition of "large in size" is :smile: . For example `Tuba` dev complains that including this made that app go up in size over 5 times.

>  legal issues like patented codecs

Personally, I do not want any web scrapping code in Clapper. For this reason this loadable at runtime extension system was made. It also benefits other apps that want to use it and if someone has a problem with web scrapping overall, its `yt-dlp` that does this, not Clapper or other app, so this makes our apps safe as far as "legal issues" go.

> complicated to maintain with the app

It is complicated to maintain with apps, because `yt-dlp` is updated frequently and since its used here as a python lib in a plugin, chances are that some newer release might change something, requiring a patch version of the plugin. In this case data checker would do PR with a patch for each app rebuilding it (if modules from here are part of their manifests). Due to constant updates of `yt-dlp`, the more apps would build this as part of them the more work for Flathub infra would be, as it would be rebuilding their whole apps instead. Also, if things change in `yt-dlp` and some compat between it and plugin needs adjusting, it can be reported and done in a single place, otherwise app devs would have more work retesting their apps with each update (possibly reverting update later on, and rebuilding again).

Currently, with reasons above this seems like win-win for everyone. The only concerns I have is what the other reviewer mentioned. That is: compatibility across base runtimes. I guess this would have to go into `branch\24.08` for this? The Flathub docs say, its not allowed to have these branches for applications, but its "fine" for extensions (another reason for separate manifest, I guess). My question is: how this works? Can I freely make these branches and infa will auto build and publish from these? Or do I have to request for adding each branch? (I couldn't find this info in docs).

GeopJr commented:
> Tuba dev complains that including this made that app go up in size over 5 times.

Tuba is *not* primarily a video player, so the only way I'm willing to bundle the enhancers (specifically yt-dlp, as clapper-enhancers is relatively small) is through an extension, either this or a tuba one.

Even outside of Clapper, a yt-dlp extension would benefit many apps, both on size and maintenance. From a quick search I can see apps like GridPlayer and Hidamari depend on much older yt-dlp versions that are likely not working anymore.

bbhtt commented:
Is clapper enhancers supposed to be used by apps other than clapper? Or do they need yt-dlp only?

Rafostar commented:
> Is clapper enhancers supposed to be used by apps other than clapper? Or do they need yt-dlp only?

This PR aims to ship "enhancers" with a corresponding `yt-dlp` to be imported by plugin module for apps using Clapper library. Apps that I know of, that currently want to use it (clapper enhancers) are: Clapper, Pipeline (tubefeeder), NewsFlash and Tuba.

Rafostar commented:
@bbhtt If I exported enhancers as a "bundle" from Clapper application manifest instead, would that be preferred? Also, would that allow other app to install that "bundled extension" without pulling whole Clapper application with it?

bbhtt commented:
it will work but depends on what you want to do. If this is a one off extension you can use bundle, but if there's going to more extensions submitted by others you need an extension point.


Also app extensions don't need to have separate branches for separate runtimes, it follows the app - you don't have multiple branches for the app.

Rafostar commented:
> Also app extensions don't need to have separate branches for separate runtimes, it follows the app - you don't have multiple branches for the app

I found this manifest as a bundle example: [SonoBus](https://github.com/flathub/net.sonobus.SonoBus/blob/060cdf42f9d1e7594b2be76ee1374d0df2fecebd/net.sonobus.SonoBus.yaml#L20C3-L20C45)

It seems I should be able to specify version like `24.08` from within app manifest and when I run `flatpak install org.freedesktop.LinuxAudio.Plugins.SonoBus` from command line it suggests multiple (previous versions). So if I understand correctly, after version update of bundle, old versions are not deleted (compat kept) but no maintaining (backporting) multiple versions at once? It should be good enough, I think.

bbhtt commented:
That's not an app extension, the linuxaudio plugins are created against the separate base-extension which is branched according to the runtime.

bbhtt commented:
> So if I understand correctly, after version update of bundle, old versions are not deleted (compat kept) but no maintaining (backporting) multiple versions at once? It should be good enough, I think.

You can't specify arbitrary branches if bundling. The branch of the app `//stable` has to match the extension's too.

Rafostar commented:
Thanks for explaining 👍 

If bundle is limited to stable, then any potential app that would use this and stop being updated/maintained might run eventually into versions conflict that other reviewer mentioned. In this case, this PR hare would avoid it (assuming it can have branches). Can I freely make these branches and infa will auto build and publish from these? Or do I have to request for adding each branch? (I couldn't find this info in docs).

bbhtt commented:
Yea you can create branches freely but generally it would if you made some API/ABI break in clapper, not over runtime versions.

If apps stop being maintained then it's not your concern to keep supporting super old versions. It's not realistic to maintain N different branches.

Usually people maintain 1 or 2.

Rafostar commented:
> If apps stop being maintained then it's not your concern to keep supporting old versions and old runtimes.
> I'm also not sure if creating branches is worth it for these two components here.

Well, if that's the case, then I would simply "bundle" this from Clapper manifest if you are OK with it. Without doing branches there doesn't seem any benefit in maintaining yet another repo.

bbhtt commented:
Yea sure, bundle requires an exception to the linter https://docs.flathub.org/docs/for-app-authors/linter#manifest-has-bundled-extension. You can get your PR for the extension ready, then open an exception request.

Rafostar commented:
Since contents of this PR ultimately ended up as a bundle, this here is no longer needed, so closing. Thank you for your help. :+1:
